lib/sphinx/integration/extensions/riddle/client.rb

```diff
@@ -38,46 +38,62 @@ def connect_with_pooling
-                    [0, messages.length].pack('NN') + message, 0
+              begin
+                case command
+                when :search
+                  if ::Riddle::Client::Versions[command] >= 0x118
+                    socket.send request_header(command, message.length) +
+                      [0, messages.length].pack('NN') + message, 0
+                  else
+                    socket.send request_header(command, message.length) +
+                      [messages.length].pack('N') + message, 0
+                  end
+                when :status
+                  socket.send request_header(command, message.length), 0
                 else
-                  socket.send request_header(command, message.length) +
-                    [messages.length].pack('N') + message, 0
+                  socket.send request_header(command, message.length) + message, 0
                 end
-              when :status
-                socket.send request_header(command, message.length), 0
-              else
-                socket.send request_header(command, message.length) + message, 0
-              end
 
-              header = socket.recv(8)
-              status, version, length = header.unpack('n2N')
-
-              while response.length < (length || 0)
-                part = socket.recv(length - response.length)
+                # Receive header
+                if read_timeout.nil?
+                  header = socket.recv(HEADER_LENGTH)
+                else
+                  nonblock_read!(socket, HEADER_LENGTH, header, read_timeout)
+                end
+                status, version, length = header.unpack('n2N')
+                length = length.to_i
 
-                # will return 0 bytes if remote side closed TCP connection, e.g, searchd segfaulted.
-                break if part.length == 0 && socket.is_a?(TCPSocket)
+                # Если вернулся ответ Retry, то нужно попробовать
+                # отправить запрос на другую ноду, если имеется таковая
+                # дальнейшее чтение из сокета ненужно, закроем его выше
+                if ::Riddle::Client::Statuses[:retry] == status
+                  raise ::Riddle::ResponseError, 'Searchd responded with retry error'
+                end
 
-                response << part if part
-              end
+                # Receive body
+                if read_timeout.nil?
+                  block_read(socket, length, response)
+                else
+                  nonblock_read!(socket, length, response, read_timeout)
+                end
 
-              if response.empty? || response.length != length
-                raise ::Riddle::ResponseError, "No response from searchd (status: #{status}, version: #{version})"
+                if response.empty? || response.bytesize != length
+                  raise ::Riddle::ResponseError, "No response from searchd (status: #{status}, version: #{version})"
+                end
+              rescue Exception
```

Comment:
Lint/RescueException: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?